README.md

```diff
@@ -574,6 +574,11 @@ it matters which key we use, really, as long as the transaction succeeds. I
 believe the goal of this step is just to get the target acceptor to receive
 _any_ ballot number for _any_ key.
 
+> In the paper cluster membership change is described before an introduction of
+> multiple keys, so in multi key case we need to execute identity state transition
+> for **every** key. Thus the acceptors should expose API to fetch a list of existing
+> keys.
```

Comment:
Whoa, really? Is there no hack to get around this one? That's gonna be impractical :(

Reply:
The straightforward approach requires to move R(2F+2) records where R is the number of records. I guess, instead of re-scan we can ask F+1 nodes to backup its state into a new node (and keeping records with max ballot), in this case we need to move R(F+1) records.

If we have a background catch-up process and we know that a node knows all data up to recent moment then we can initialize a new node moving only (R-r) + r(F+1) records where r << R.

Reply:
If each multi-key acceptor tracked the max ballot it accepted for _any_ key as `B_max`, and during conf change we call some method on each acceptor which updates the accepted ballot number for _every_ key to `B_max+1`, would that be sufficient?

Reply:
No, as I understand we need to move data. Think of it as a catch-up, when we add a new node to the system we want to replicate existing data there otherwise replication factor drops below the quorum size we may violate consistency.

The same is true in log based systems, adding a new node requires moving the whole dataset but it has lower amplification factor there.

Reply:
OK, I think I have a strategy, I will write it up soon.